### PR TITLE
Add dedicated North Arrow creation action

### LINKED ISSUE #30162: Adding a north arrow is too hard
At the moment the logic to add a north arrow to a map is:

- Add picture
- Find Item Properties Tab
- Expand Search Directories
- Skip the first 30 or so icons
- Find a north arrow
- Update SVG parameters for style

I know that north arrows are just images however I think we can do something better here around this in order to make the experience better for the users.

- New North Arrow button
- Opens SVG properties
- Auto close main properties
- Auto expand Search Directories
- Filter the SVG list if we can to just show the north arrows
- Auto enable Sync with map" for rotation.

### PR DESCRIPTION
**Pull request by nyalldawson**
This is a shortcut to adding a picture item, setting it to a north arrow picture, and linking it with a map. The end result is identical, but it's much easier for new users to understand if we expose it as an explicit "North Arrow" item.

Even experienced users will likely appreciate the improved workflow, including automatically linking the picture rotation to a sensible default map choice (if a map is selected, it's used. If not, the topmost map item under the newly drawn north arrow is used. If there's none, the layout's 'reference map' (or biggest map) is used as a fallback)

Fixes #30162

Peek 2019-06-12 14-25

### CODE REVIEW

**Comment by NathanW2** [comment]
image

Love it! Thanks heaps.

**Comment by nirvn** [comment]
Can't figure out whether you were trying to type lotus or layout in your commit message ;)

I love it!

**Comment by nirvn** [comment]
@nyalldawson , code looks good.

**Comment by roya0045** [comment]
Great! Is there a way to change the default svg that will be used? 
Or will that be an optional step after the arrow is created?

**Comment by Gustry** [comment]
Nice addition !
Side question, is-it possible to customize the default picture which is used with an settings in the INI file? This would be nice to have in an organization.

**Comment by nyalldawson** [comment]
@Gustry good point, added

@roya0045 After creation the north arrows are identical to standard picture items, so you can go into their properties and change the picture to anything you want (including the other north arrows provided by default in a qgis install)

**Comment by roya0045** [comment]
> @Gustry good point, added
> 
> @roya0045 After creation the north arrows are identical to standard picture items, so you can go into their properties and change the picture to anything you want (including the other north arrows provided by default in a qgis install)

The point raised by Gustry was what I had in mind. I was just thinking from a new user perspective if they wanted to add a new arrow and if they always had to change it afterward to another design, it might be an extra convenience to go through some extra step in order to not have to always change the design of the arrow.

File: src/app/layout/qgslayoutvaliditychecks.cpp [hunk 1]
```diff
@@ -73,6 +73,62 @@ QList<QgsValidityCheckResult> QgsLayoutScaleBarValidityCheck::runCheck( const Qg
...
+  if ( !layoutContext )
+    return false;
+
+  QList< QgsLayoutItemPicture * > pictureItems;
+  layoutContext->layout->layoutItems( pictureItems );
+  for ( QgsLayoutItemPicture *picture : qgis::as_const( pictureItems ) )
+  {
+    // look for pictures which use the default north arrow svg, but aren't actually linked to maps.
+    // alternatively identify them by looking for the default "North Arrow" string in their id
+    if ( !picture->linkedMap() && ( picture->picturePath() == QStringLiteral( ":/images/north_arrows/layout_default_north_arrow.svg" )
```
Comment (Gustry) [inline src/app/layout/qgslayoutvaliditychecks.cpp]: Update here too with the qsettings?

File: resources/qgis_global_settings.ini [hunk 2]
```diff
@@ -99,6 +99,12 @@ PostgreSQL\default_timeout=30
 # cause performance issues, as the records must all be loaded from the related table on display.
 maxEntriesRelationWidget=100
 
+# Path to default image to use for layout north arrows
+LayoutDesigner\defaultNorthArrow=:/images/north_arrows/layout_default_north_arrow.svg
```
Comment (Gustry) [inline resources/qgis_global_settings.ini]: So can we can give a HTTP URL or a local filepath, without using a QRC file? Nice Thanks!

**Review by Gustry — APPROVED** [review]
Thanks!

File: src/app/layout/qgslayoutvaliditychecks.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/app/layout/qgslayoutvaliditychecks.cpp]: Thanks, fixed

File: resources/qgis_global_settings.ini (hunk 2 above)
Comment (nyalldawson) [reply resources/qgis_global_settings.ini]: Yes, fixed. Even base64 embedded north arrows are permitted :D

**Comment by nyalldawson** [comment]
Related mailing list discussion https://lists.osgeo.org/pipermail/qgis-user/2019-February/042354.html

**Comment by nirvn** [comment]
This looks good to merge. Nice UX boost when it comes to layout creation.

**Comment by nyalldawson** [comment]
So merge for 3.8? or defer to 3.10?

**Comment by nirvn** [comment]
IMO, merge for 3.8. This is a UX fix more than anything else. Plus, the way it's implemented by using a preexisting picture widget is not really introducing anything new.

The refined map association logic is also most welcome for the scale bar and legend.

**Pull request merged**